x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/inference/Rerank.java

```diff
@@ -0,0 +1,169 @@
+        out.writeCollection(rerankFields());
+        out.writeNamedWriteable(scoreAttribute);
+    }
+
+    public Expression queryText() {
+        return queryText;
+    }
+
+    public List<Alias> rerankFields() {
+        return rerankFields;
+    }
+
+    public Attribute scoreAttribute() {
+        return scoreAttribute;
+    }
+
+    @Override
+    public TaskType taskType() {
+        return TaskType.RERANK;
+    }
+
+    @Override
+    public Rerank withInferenceResolutionError(String inferenceId, String error) {
+        Expression newInferenceId = new UnresolvedAttribute(inferenceId().source(), inferenceId, error);
+        return new Rerank(source(), child(), newInferenceId, queryText, rerankFields, scoreAttribute);
+    }
+
+    public Rerank withRerankFields(List<Alias> newRerankFields) {
+        return new Rerank(source(), child(), inferenceId(), queryText, newRerankFields, scoreAttribute);
+    }
+
+    public Rerank withScoreAttribute(Attribute newScoreAttribute) {
+        return new Rerank(source(), child(), inferenceId(), queryText, rerankFields, newScoreAttribute);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public UnaryPlan replaceChild(LogicalPlan newChild) {
+        return new Rerank(source(), newChild, inferenceId(), queryText, rerankFields, scoreAttribute);
+    }
+
+    @Override
+    protected AttributeSet computeReferences() {
+        AttributeSet refs = computeReferences(rerankFields);
+        refs.add(scoreAttribute);
+        return refs;
+    }
+
+    public static AttributeSet computeReferences(List<Alias> fields) {
+        AttributeSet generated = new AttributeSet(asAttributes(fields));
+        return Expressions.references(fields).subtract(generated);
+    }
+
+    @Override
+    public boolean expressionsResolved() {
+        return super.expressionsResolved() && queryText.resolved() && Resolvables.resolved(rerankFields) && scoreAttribute.resolved();
+    }
+
+    @Override
+    protected NodeInfo<? extends LogicalPlan> info() {
+        return NodeInfo.create(this, Rerank::new, child(), inferenceId(), queryText, rerankFields, scoreAttribute);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        if (super.equals(o) == false) return false;
+        Rerank rerank = (Rerank) o;
+        return Objects.equals(queryText, rerank.queryText)
+            && Objects.equals(rerankFields, rerank.rerankFields)
+            && Objects.equals(scoreAttribute, rerank.scoreAttribute);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), queryText, rerankFields, scoreAttribute);
+    }
+
+    @Override
+    public LogicalPlan surrogate() {
+        Order sortOrder = new Order(source(), scoreAttribute, Order.OrderDirection.DESC, Order.NullsPosition.ANY);
```

Comment:
ℹ️ `RERANK` is a surrogate plan, and a `| SORT BY _score DESC` is automatically added after it.